iceoryx_posh/test/moduletests/test_roudi_iceoryx_roudi_app.cpp

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2021 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_posh/internal/log/posh_logging.hpp"
+#include "iceoryx_posh/roudi/iceoryx_roudi_app.hpp"
+#include "iceoryx_posh/roudi/roudi_cmd_line_parser_config_file_option.hpp"
+#include "iceoryx_posh/roudi/roudi_config_toml_file_provider.hpp"
+#include "iceoryx_utils/platform/getopt.hpp"
+
+#include "test.hpp"
+
+using namespace ::testing;
+using ::testing::Return;
+
+using iox::roudi::IceOryxRouDiApp;
+using namespace iox::config;
+
+namespace iox
+{
+namespace test
+{
+class IceoryxRoudiApp_Child : public IceOryxRouDiApp
+{
+  public:
+    IceoryxRoudiApp_Child(const config::CmdLineArgs_t& cmdLineArgs, const RouDiConfig_t& roudiConfig)
+        : IceOryxRouDiApp(cmdLineArgs, roudiConfig)
+    {
+    }
+
+    bool getVariableRun()
+    {
+        return m_run;
+    }
+
+    iox::log::LogLevel getLogLevel()
+    {
+        return m_logLevel;
+    }
+
+    roudi::MonitoringMode getMonitoringMode()
+    {
+        return m_monitoringMode;
+    }
+
+    void setVariableRun(bool condition)
+    {
+        m_run = condition;
+    }
+
+    uint8_t callRun() noexcept
+    {
+        return run();
+    }
+};
+
+/// @brief Test goal: This file tests class roudi app
+class IceoryxRoudiApp_test : public Test
+{
+  public:
+    CmdLineParserConfigFileOption cmdLineParser;
+
+    void TearDown()
+    {
+        // Reset optind to be able to parse again
+        optind = 0;
+    };
+};
+
+TEST_F(IceoryxRoudiApp_test, VerifyConstructorIsSuccessfull)
+{
+    constexpr uint8_t numberOfArgs{1U};
+    char* args[numberOfArgs];
+    char appName[] = "./foo";
+    args[0] = &appName[0];
+
+    auto cmdLineArgs = cmdLineParser.parse(numberOfArgs, args);
+
+    IceoryxRoudiApp_Child roudi(cmdLineArgs.value(), iox::RouDiConfig_t().setDefaults());
+
+    EXPECT_TRUE(roudi.getVariableRun());
+    EXPECT_THAT(roudi.getLogLevel(), Eq(iox::log::LogLevel::kWarn));
+    EXPECT_THAT(roudi.getMonitoringMode(), Eq(roudi::MonitoringMode::ON));
+}
+
+TEST_F(IceoryxRoudiApp_test, CreateTwoRoudiAppIsSuccessfull)
+{
+    constexpr uint8_t numberOfArgs{1U};
+    char* args[numberOfArgs];
+    char appName[] = "./foo";
+    args[0] = &appName[0];
+
+    auto cmdLineArgs = cmdLineParser.parse(numberOfArgs, args);
+
+    IceoryxRoudiApp_Child roudi(cmdLineArgs.value(), iox::RouDiConfig_t().setDefaults());
```

Comment:
I think we do not want this kind of behavior or is there a use case that the `RoudiApp` should be instantiated twice?!

Reply:
I guess this come from the current implementation. The `RouDiLock` is located in `IceOryxRouDiComponents`.
The Components are only instantiated when you call `IceOryxRouDiApp::run()`. Here in this case only the class is instantiated... Maybe the lock should be relocated to fire earlier.

Reply:
Well, nothing bad happens until `run` is called, so this is no problem. What's instantiated is just a small shell around some configuration data, the actual roudi is created in `run`

Reply:
Yeah, the lock's description is `Locks the socket for preventing multiple start of RouDi`, so, is the intention just to prevent `run` or to instantiate it? In case of `run`-only another test for that is missing?!

Reply:
from my point of view it should only be `run`, else you cannot execute `./iox-roudi -v` anymore when RouDi is running. All the stuff that can break if tow instances are running happens in the `run` method

Reply:
Makes sense and I'm fine with that. So, is the PR complete now or should a 2nd run be tested?

Reply:
As discussed in the dev meeting, a test for a 2nd run will be done in #621.